kasa/tapo/tapodevice.py

```diff
@@ -51,19 +54,51 @@ async def update(self, update_children: bool = True):
         _LOGGER.debug("Got an update: %s", self._data)
 
     @property
+    @requires_update
     def sys_info(self) -> Dict[str, Any]:
         """Returns the device info."""
-        return self._info
+        return self._info  # type: ignore
 
     @property
+    @requires_discovery_or_update
     def model(self) -> str:
         """Returns the device model."""
-        return str(self._info.get("model"))
+        if self._info:
+            return str(self._info.get("model"))
+        else:
+            return str(self._discovery_info.get("device_model"))  # type: ignore
 
     @property
+    @requires_discovery_or_update
     def alias(self) -> str:
         """Returns the device alias or nickname."""
-        return base64.b64decode(str(self._info.get("nickname"))).decode()
+        if self._info:
+            return base64.b64decode(str(self._info.get("nickname"))).decode()
+        else:
+            return str(self._discovery_info.get("device_type"))  # type: ignore
+
+    @property
+    @requires_discovery_or_update
+    def mac(self) -> str:
+        """Return mac address.
+
+        :return: mac address in hexadecimal with colons, e.g. 01:23:45:67:89:ab
+        """
+        if self._info:
+            mac = str(self._info.get("mac")).replace("-", ":")
+        else:
+            mac = str(self._discovery_info.get("mac")).replace("-", ":")  # type: ignore
```

Comment:
We do not definitely want to do this, but instead (probably) feed `_info` with the discovery information. Basically, we want one method that takes the discovery information and constructs a data structure that looks like regular `_info`.

This way we have no branching/differential handling inside the properties, which makes it much, much easier to write tests and avoid any potential mistakes.

Reply:
Fixed